Laravel bootstrap

ThibaudDauce opened the pull request:
Right now, the extension doesn't support loading Laravel ServiceProvider for features like:
- Views namespace https://laravel.com/docs/8.x/packages#views (DDD type of application with views in different folders for each domains)
- Custom directives
- Components?

This PR add support to bootstrap the Laravel container and fetch object from Laravel instead of PHPStan. The objects will be fully registered with all custom stuff inside them (view namespaces, custom directives…).

Right now, I just tried to support custom view namespaces (it's working on my application, the extension is now loading all my views). It's really easy to add support for Blade directive too I think (juste swap the PHPStan container for the BladeCompiler for the configured Laravel one).

I put a fallback to the PHPStan container because for the tests with don't have access to Laravel (the bootstrap file doesn't exists). If we continue down this path (and I think we should load Laravel to allow for complete checks), we need to find a way to test these features (maybe [Orchestral/testbench](https://github.com/orchestral/testbench) but not sure if it's working for a PHPStan extension…). Or a fully laravel/laravel application inside a subfolder?

I'm not sure that the `templatesPaths` parameter is require if we have the bootstrap file because all information is present inside `config/view.php` and since Laravel is fully loaded, the config is loaded too.

For the loading part, I was inspired by the `CreatesApplication` trait for testing inside laravel/laravel https://github.com/laravel/laravel/blob/8.x/tests/CreatesApplication.php

ThibaudDauce committed:
Bootstrap the Laravel application to allow namespaced views

canvural commented:
Hi,

I really don't want to boot Laravel. I have experience from Larastan, and it causes too much trouble. Also I want people to be able to use this rule even if they don't have full Laravel app (requiring just `illuminate/view`). Another reason is that it feels wrong to do this when this should be static analysis.

Also the things you wrote seems to be solvable without it.

> Views namespace

Isn't this solvable with adding the paths to `templatePaths` config array?

> Custom directives

I already have an idea how to solve this. By requiring users to create a bootstrap file with the custom blade directives in it. Yeah, sounds like too much work for the user. That's why I'll keep thinking.

> Components

This is already [solved](https://github.com/canvural/phpstan-blade-rule/blob/main/src/Compiler/BladeToPHPCompiler.php#L173-L192). I just need to do more testing and write documentation about how to use it.

ThibaudDauce commented:
> I really don't want to boot Laravel. I have experience from Larastan, and it causes too much trouble. Also I want people to be able to use this rule even if they don't have full Laravel app (requiring just illuminate/view). Another reason is that it feels wrong to do this when this should be static analysis.

With my solution, the extension is working without Laravel because if the Laravel bootstsrap file is not present (or if you want, it can be an option in the config to specify the bootstrap file), the code fallback to PHPStan container and sensible defaults for views, directives and components…

I think for my use case, tens of domains with custom views namespaces and a lot of components registered with namespaces too, without booting the framework the extension will be hard to add to my project.

> Isn't this solvable with adding the paths to templatePaths config array?

No because the `loadViewsFrom()` Laravel function accept a prefix, and the views are loaded by using a syntax: `Invoice::index` where `Invoice::` is the prefix. (that's what I mean in another PR about using the "real" view name :-) )

> I already have an idea how to solve this. By requiring users to create a bootstrap file with the custom blade directives in it. Yeah, sounds like too much work for the user. That's why I'll keep thinking.

Yes I think it's error prone (and a lot of work) to let the user manually manage a bootstrap file. And it may be impossible to express everything that is possible with services providers.

> Components

I saw what you did for the components, it's cool if it's working but same thing that for Blade directives, you require the user to manually list all it's component: lot of work and error prone.

canvural commented:
Hi,

I thought about this again. But sorry, I want the analysis to remain static. With enough time we can find solutions to those problems I believe.

ThibaudDauce commented:
Hi,

I understand your view, I will look into your extension in the future when view namespacing is supported because since almost all my views are under a namespace I cannot test without it! :-)

canvural commented:
@ThibaudDauce Maybe I'm missing something, but looks like supporting view namespacing seems really easy.

So for example if you call `view('pagination::tailwind')` view will be in `vendor/laravel/framework/src/Illuminate/Pagination/resources/views` Looks like we can always drop the `xxx::` part from the view, and if the user added the vendor path to `templatePaths` array, it'll find the view.

ThibaudDauce commented:
Hi,

My views namespaces are dynamically generated. When I create a new domain folder in my application, if there is a `views` subfolder, the views are loaded with the domain prefix (I have the same thing for `routes` subfolder, `migrations` subfolder…).

This is a pretty niche use-case I admit :-D And that's why I perfectly understand why you don't want to load Laravel inside your extension.

I started working on a fork with Laravel bootstrap support this morning and I've got everything working. I may publish the code after some cleanups if you want to check some of my improvements ideas:
- I fixed some bug with `view('foo', ['bar' => $this])` not working
- I fixed a bug (not sure if you have the same one) with nested `@foreach` where PHPStan doesn't understand the `$__currentLoopData` variable (I rewrite the node to remove the `$__currentLoopData` variable and put the real variable directly inside the `foreach()` (replace `$__currentLoopData = $foo; foreach ($__currentLoopData as …)` by `foreach ($foo as …)`)
- I remove the Blade formatter for a simpler approach with (but it's maybe not a good idea to put something that is not a file path inside the file attribute :-) )
```
$errors[] = RuleErrorBuilder::message($raw_error->getMessage())
                ->file("View {$matches['template_name']} from {$file}:{$node->getLine()}")
                ->line($matches['line'])
                ->build();
```
- I add the `$errors` variable with `$variables_and_types[] = new VariableAndType('errors', new ObjectType(ViewErrorBag::class));`

I still have 763 errors so I may find other stuff :-D

canvural commented:
> My views namespaces are dynamically generated.

Ok, but still the technique I wrote would work I guess. You just need to add that new domain view folder to `templatePaths`

> I may publish the code after some cleanups if you want to check some of my improvements ideas

Feel free to send a PR for bugfixes 👍🏽
